crates/wasmtime/src/runtime/vm/const_expr.rs

```diff
@@ -43,6 +47,116 @@ impl<'a, 'b> ConstEvalContext<'a, 'b> {
             self.instance.get_func_ref(index).unwrap().cast(),
         ))
     }
+
+    fn struct_fields_len(&self, struct_type_index: ModuleInternedTypeIndex) -> usize {
+        let module = self
+            .instance
+            .runtime_module()
+            .expect("should never be allocating a struct type defined in a dummy module");
+
+        let struct_ty = match &module.types()[struct_type_index].composite_type {
+            WasmCompositeType::Struct(s) => s,
+            _ => unreachable!(),
+        };
+
+        struct_ty.fields.len()
+    }
+
+    /// Safety: field values must be of the correct types.
+    unsafe fn struct_new(
+        &mut self,
+        struct_type_index: ModuleInternedTypeIndex,
+        fields: &[ValRaw],
+    ) -> Result<ValRaw> {
+        let module = self
+            .instance
+            .runtime_module()
+            .expect("should never be allocating a struct type defined in a dummy module");
+        let shared_ty = module
+            .signatures()
+            .shared_type(struct_type_index)
+            .expect("should have an engine type for module type");
+
+        let store = unsafe { (*self.instance.store()).store_opaque_mut() };
+
+        let struct_ty = StructType::from_shared_type_index(store.engine(), shared_ty);
+        let fields = fields
+            .iter()
+            .zip(struct_ty.fields())
+            .map(|(raw, ty)| {
+                let ty = ty.element_type().unpack();
+                let mut store = AutoAssertNoGc::new(store);
+                Val::_from_raw(&mut store, *raw, ty)
+            })
+            .collect::<Vec<_>>();
+
+        let allocator = StructRefPre::_new(store, struct_ty);
+        let mut store = OpaqueRootScope::new(store);
+
+        // TODO: if this fails with `GcHeapOutOfMemory<()>`, then we should do a
+        // GC and try to allocate again, hoping that the GC freed up
+        // space. However, we would need to choose between sync and async GC,
+        // and we can't use `AsyncCx::block_on` here to fake it because we
+        // aren't guaranteed to be running on a fiber yet, so we would be forced
+        // to create sync and async versions of every function up to at least
+        // `ConstExprEvaluator::eval`.
+        //
+        // In the meantime, just give up early without trying a GC if allocation
+        // fails.
+        let struct_ref = StructRef::_new(&mut store, &allocator, &fields)?;
+
+        let mut store = AutoAssertNoGc::new(&mut store);
```

Comment:
This has me thinking, isn't rooting a problem for the entire duration of the const eval? For example we've just got these `ValRaw::anyref` values floating around internally on the const eval stack, so if a GC were to happen in `StructRef::_new` above, for example, it wouldn't be able to find those roots?

Should `AutoAssertNoGc` be wrapped around the entire const eval operation?